File: src/input/commands.ts
/**
 * Command parser for k9s-style : commands
 */

export interface ParsedCommand {
  command: string;
  args: string[];
}

// Command aliases (short form -> canonical form)
const COMMAND_ALIASES: Record<string, string> = {
  wf: "workflows",
  sch: "schedules",
  tq: "taskqueues",
  ns: "namespace",
  q: "quit",
  h: "help",
};

// Valid commands
const VALID_COMMANDS = new Set([
  "workflows",
  "schedules",
  "taskqueues",
  "namespace",
  "quit",
  "help",
]);

/**
 * Parse a command string into command and arguments
 * @param input - Raw input string (without leading :)
 * @returns Parsed command or null if invalid
 */
export function parseCommand(input: string): ParsedCommand | null {
  const trimmed = input.trim().toLowerCase();
  if (!trimmed) return null;

  const parts = trimmed.split(/\s+/);
  const rawCommand = parts[0] ?? "";
  const args = parts.slice(1);

  // Resolve alias to canonical command
  const command = COMMAND_ALIASES[rawCommand] ?? rawCommand;

  // Validate command
  if (!VALID_COMMANDS.has(command)) {
    return null;
  }

  return { command, args };
}

/**
 * Get command suggestions for auto-complete
 * @param partial - Partial command string
 * @returns Array of matching command names
 */
export function getCommandSuggestions(partial: string): string[] {
  const lower = partial.toLowerCase();
  const suggestions: string[] = [];

  // Check aliases
  for (const [alias, command] of Object.entries(COMMAND_ALIASES)) {
    if (alias.startsWith(lower) || command.startsWith(lower)) {
      if (!suggestions.includes(alias)) {
        suggestions.push(alias);
      }
    }
  }

  // Check full command names
  for (const command of VALID_COMMANDS) {
    if (command.startsWith(lower) && !suggestions.includes(command)) {
      suggestions.push(command);
    }
  }

  return suggestions.sort();
}

/**
 * Get help text for a command
 */
export function getCommandHelp(command: string): string {
  const resolved = COMMAND_ALIASES[command] ?? command;

  switch (resolved) {
    case "workflows":
      return "Switch to workflows view";
    case "schedules":
      return "Switch to schedules view";
    case "taskqueues":
      return "Switch to task queues view";
    case "namespace":
      return "Switch namespace (use :ns <name> or :ns to open selector)";
    case "quit":
      return "Quit the application";
    case "help":
      return "Show help";
    default:
      return "";
  }
}

File: src/input/commands.rs
pub struct CommandDef {
    pub name: &'static str,
    pub aliases: &'static [&'static str],
    pub description: &'static str,
}

pub static COMMANDS: &[CommandDef] = &[
    CommandDef {
        name: "workflows",
        aliases: &["wf"],
        description: "Switch to workflows view",
    },
    CommandDef {
        name: "schedules",
        aliases: &["sch"],
        description: "Switch to schedules view",
    },
    CommandDef {
        name: "namespace",
        aliases: &["ns"],
        description: "Switch namespace (e.g. :ns production)",
    },
    CommandDef {
        name: "signal",
        aliases: &["sig"],
        description: "Signal workflow (e.g. :signal my-signal {\"key\":\"val\"})",
    },
    CommandDef {
        name: "quit",
        aliases: &["q"],
        description: "Quit t9s",
    },
    CommandDef {
        name: "help",
        aliases: &["h"],
        description: "Show help",
    },
];

pub fn matching_commands(input: &str) -> Vec<&'static CommandDef> {
    let input_lower = input.to_lowercase();
    COMMANDS
        .iter()
        .filter(|cmd| {
            cmd.name.starts_with(&input_lower)
                || cmd.aliases.iter().any(|a| a.starts_with(&input_lower))
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_matching_commands() {
        assert_eq!(matching_commands("w").len(), 1);
        assert_eq!(matching_commands("w")[0].name, "workflows");

        assert_eq!(matching_commands("wf").len(), 1);
        assert_eq!(matching_commands("wf")[0].name, "workflows");

        assert_eq!(matching_commands("s").len(), 2); // schedules + signal
        assert_eq!(matching_commands("sch").len(), 1);
        assert_eq!(matching_commands("sch")[0].name, "schedules");

        assert_eq!(matching_commands("sig").len(), 1);
        assert_eq!(matching_commands("sig")[0].name, "signal");

        assert_eq!(matching_commands("q").len(), 1);
        assert_eq!(matching_commands("q")[0].name, "quit");

        assert!(matching_commands("xyz").is_empty());
    }
}
